notebooks/03-query/11-joins.ipynb

```diff
@@ -0,0 +1,2475 @@
+       " {'query_type': 'simple_prejoin',\n",
+       "  'results': 'Results = avg:128.40235 ms   min:122.407 ms  max:176.861 ms'},\n",
+       " {'query_type': 'subquery_join',\n",
+       "  'results': 'Results = avg:2959.44585 ms   min:2841.148 ms  max:3093.195 ms'},\n",
+       " {'query_type': 'subquery_join_lookup',\n",
+       "  'results': 'Results = avg:550.8509 ms   min:519.703 ms  max:608.721 ms'},\n",
+       " {'query_type': 'subquery_prejoin',\n",
+       "  'results': 'Results = avg:352.1546 ms   min:327.998 ms  max:471.693 ms'}]"
+      ]
+     },
+     "execution_count": 22,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# review the results\n",
+    "sorted_list = sorted(join_results, key=lambda x: x['query_type'])\n",
+    "sorted_list"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "66fd0a23-cc14-4537-8965-5dab8eeb1b46",
+   "metadata": {},
+   "source": [
+    "In these results you can see the progression of response times. \n",
+    "- The use of lookups improves performance over joining to another datasource, but it requires enough Heap to fit all lookups on all historical and ingestion tasks (Peon) JVMs.\n",
+    "- The pre-joined table performs even better and has the added advantage of not requiring any additional Heap.\n",
+    "\n",
+    "If the data being joined changes over time and you wish to query with the latest values, consider a lookup. If the data does not change or the change is not desirable, then pre-join the data. The example of `age_group` as the joined property works well in the pre-join strategy because the user's age_group is important at the time they clicked, not when looking at old click rows based on their current age_group."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "46e2e6f6-86e1-48dd-a125-6c1920173d74",
+   "metadata": {},
+   "source": [
+    "## Asynchronous Queries on MSQ Engine\n",
+    "\n",
+    "The MSQ (Multi-stage Query Engine) engine works by shuffling rows between stages. All stages execute in tasks in parallel as much as they can.\n",
+    "![](assets/msq-engine.png)\n",
+    "\n",
+    "Typically, as you did in the ingestion above, the MSQ engine is used to ingest external data, but it can also be used to query Druid datasources directly. It has the advantage of being able to do either `broadcast` or `sortMerge` joins. \n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "e8b27edf-90f2-4f69-90e1-c5db281f2934",
+   "metadata": {},
+   "source": [
+    "### Broadcast joins in MSQ\n",
+    "Broadcast joins execute similarly to the gather/scatter approach in that they send the whole right-hand dataset to all workers, but MSQ runs asynchronously and does not have the `maxSubQueryRows` limit. It will broadcast and load the whole right hand side table on all query_worker tasks.\n",
+    "![](assets/msq-broadcast.png)\n",
```

Comment:
Documentation.

Reply:
Where? I don't see explanations about the MSQ engine processing model or how broadcast or sortMerge work, just how to specify them.... I think this is useful for people to understand.